prompt2model/model_trainer/qlora_trainer.py

```diff
@@ -0,0 +1,248 @@
+
+    def __init__(self, val_data: datasets.Dataset, eval_tokenizer) -> None:
+        """Initialize the callback with the validation data and tokenizer.
+
+        Args:
+            val_data: The val dataset. Has 2 columns: "input_col" and "output_col".
+            eval_tokenizer: The tokenizer to use for evaluation.
+
+        Note: The val_data is shuffled and a max of 100 samples are used for evaluation.
+        """
+        range_len = min(100, len(val_data))
+        self.val_data = val_data.shuffle(seed=42).select(range(range_len))
+        self.eval_tokenizer = eval_tokenizer
+        self.device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
+
+    def on_evaluate(self, args, state, control, model, metrics, **kwargs):
+        """Evaluate the model on the validation set and log the accuracy."""
+        # Evaluate the model on the validation set
+        # 1. use the val data
+        # 2. use model and tokenizer
+        # 3. get accuracy
+        val_acc = 0
+        with torch.no_grad():
+            for row in self.val_data:
+                input_text = row["input_col"]
+                target = row["output_col"]
+
+                max_gen_len = (
+                    len(self.eval_tokenizer(target, return_tensors="pt").input_ids[0])
+                    + 5
+                )
+                inpt_ids = self.eval_tokenizer(
+                    input_text, return_tensors="pt"
+                ).input_ids.to(self.device)
+                generated_ids = model.generate(
+                    input_ids=inpt_ids,
+                    max_new_tokens=max_gen_len,
+                    do_sample=False,
+                    temperature=0,
+                )
+                text = self.eval_tokenizer.decode(
+                    generated_ids[0], skip_special_tokens=True
+                )
+                text = text.replace(input_text, "").lower().strip()
+                target = target.lower().strip()
+                if target in text:
+                    val_acc += 1
+        val_acc = val_acc / len(self.val_data)
+        print(f"Validation accuracy: {val_acc}")
+        if metrics is not None:
+            metrics["eval_accuracy"] = val_acc
+
+
+class QLoraTrainer:
+    """A class for fine-tuning a model using QLora."""
+
+    def __init__(self, model_name: str, model_max_length: int) -> None:
+        """Initialize the QLoraTrainer with a model name and evaluation size."""
+        self.model_name = model_name
+        self.model_max_length = model_max_length
+        self.bnb_config = BitsAndBytesConfig(
+            load_in_4bit=True,
+            bnb_4bit_use_double_quant=True,
+            bnb_4bit_quant_type="nf4",
+            bnb_4bit_compute_dtype=torch.bfloat16,
+        )
+        self.model = AutoModelForCausalLM.from_pretrained(
+            self.model_name, device_map="auto", quantization_config=self.bnb_config
+        )
+        print("Model loaded")
```

Comment:
use `logger.info` instead of `print` (e.g. https://github.com/neulab/prompt2model/blob/67673b1725d0d95ab1b109d65c3fd755760dd45a/prompt2model/model_evaluator/seq2seq.py#L59)